File: sdk/runtime/msgpack.ts
// MessagePack's standard scalar, binary, array and string-keyed map types.
// Extension types are deliberately unsupported: Fish's protocol does not use them.
export function encodeMessagePack(value: unknown): Uint8Array<ArrayBuffer> {
  const chunks: Uint8Array[] = [];
  const encoder = new TextEncoder();
  const header = (marker: number, length: number, size: number) => {
    const bytes = new Uint8Array(1 + size);
    bytes[0] = marker;
    const view = new DataView(bytes.buffer);
    if (size === 1) view.setUint8(1, length);
    else if (size === 2) view.setUint16(1, length);
    else if (size === 4) view.setUint32(1, length);
    chunks.push(bytes);
  };
  const write = (item: unknown, depth: number): void => {
    if (depth > 64) throw new TypeError("MessagePack nesting exceeds 64 levels");
    if (item === null || typeof item === "boolean") { header(item === null ? 0xc0 : item ? 0xc3 : 0xc2, 0, 0); return; }
    if (typeof item === "number") {
      if (!Number.isFinite(item)) throw new TypeError("MessagePack numbers must be finite");
      if (Number.isInteger(item) && item >= 0 && item <= 0x7f) header(item, 0, 0);
      else if (Number.isInteger(item) && item >= -32 && item < 0) header(256 + item, 0, 0);
      else if (Number.isInteger(item) && item >= 0 && item <= 0xff) header(0xcc, item, 1);
      else if (Number.isInteger(item) && item >= 0 && item <= 0xffff) header(0xcd, item, 2);
      else if (Number.isInteger(item) && item >= 0 && item <= 0xffffffff) header(0xce, item, 4);
      else if (Number.isInteger(item) && item >= -0x80000000 && item < 0) header(0xd2, item, 4);
      else {
        const bytes = new Uint8Array(9); bytes[0] = 0xcb;
        new DataView(bytes.buffer).setFloat64(1, item); chunks.push(bytes);
      }
      return;
    }
    if (typeof item === "string") {
      const bytes = encoder.encode(item);
      if (bytes.length < 32) header(0xa0 | bytes.length, 0, 0);
      else if (bytes.length <= 0xff) header(0xd9, bytes.length, 1);
      else if (bytes.length <= 0xffff) header(0xda, bytes.length, 2);
      else header(0xdb, bytes.length, 4);
      chunks.push(bytes); return;
    }
    if (item instanceof Uint8Array) {
      if (item.length <= 0xff) header(0xc4, item.length, 1);
      else if (item.length <= 0xffff) header(0xc5, item.length, 2);
      else header(0xc6, item.length, 4);
      chunks.push(item); return;
    }
    if (Array.isArray(item)) {
      if (item.length < 16) header(0x90 | item.length, 0, 0);
      else if (item.length <= 0xffff) header(0xdc, item.length, 2);
      else header(0xdd, item.length, 4);
      for (const child of item) write(child, depth + 1);
      return;
    }
    if (typeof item === "object" && item && (Object.getPrototypeOf(item) === Object.prototype || Object.getPrototypeOf(item) === null)) {
      const entries = Object.entries(item).filter(([, child]) => child !== undefined);
      if (entries.length < 16) header(0x80 | entries.length, 0, 0);
      else if (entries.length <= 0xffff) header(0xde, entries.length, 2);
      else header(0xdf, entries.length, 4);
      for (const [key, child] of entries) { write(key, depth + 1); write(child, depth + 1); }
      return;
    }
    throw new TypeError("Unsupported MessagePack value");
  };
  write(value, 0);
  const output = new Uint8Array(chunks.reduce((size, chunk) => size + chunk.length, 0));
  let offset = 0;
  for (const chunk of chunks) { output.set(chunk, offset); offset += chunk.length; }
  return output;
}

export function decodeMessagePack(data: ArrayBuffer | ArrayBufferView): unknown {
  const bytes = data instanceof ArrayBuffer ? new Uint8Array(data) : new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
  const decoder = new TextDecoder("utf-8", { fatal: true });
  let offset = 0;
  const take = (length: number) => {
    if (length > bytes.length - offset) throw new TypeError("Truncated MessagePack value");
    const result = bytes.subarray(offset, offset + length); offset += length; return result;
  };
  const number = (length: number, signed = false, float = false): number => {
    const part = take(length); const view = new DataView(part.buffer, part.byteOffset, part.byteLength);
    if (float) return length === 4 ? view.getFloat32(0) : view.getFloat64(0);
    if (length === 1) return signed ? view.getInt8(0) : view.getUint8(0);
    if (length === 2) return signed ? view.getInt16(0) : view.getUint16(0);
    if (length === 4) return signed ? view.getInt32(0) : view.getUint32(0);
    const value = Number(signed ? view.getBigInt64(0) : view.getBigUint64(0));
    if (!Number.isSafeInteger(value)) throw new TypeError("MessagePack integer exceeds the safe integer range");
    return value;
  };
  const array = (length: number, depth: number): unknown[] => {
    if (length > bytes.length - offset) throw new TypeError("Truncated MessagePack value");
    const result: unknown[] = [];
    for (let index = 0; index < length; index++) result.push(read(depth + 1));
    return result;
  };
  const map = (length: number, depth: number): Record<string, unknown> => {
    if (length > (bytes.length - offset) / 2) throw new TypeError("Truncated MessagePack value");
    const result: Record<string, unknown> = {};
    for (let index = 0; index < length; index++) {
      const key = read(depth + 1);
      if (typeof key !== "string") throw new TypeError("MessagePack map key is not a string");
      if (Object.hasOwn(result, key)) throw new TypeError("MessagePack map contains duplicate keys");
      Object.defineProperty(result, key, { value: read(depth + 1), enumerable: true, configurable: true, writable: true });
    }
    return result;
  };
  const read = (depth: number): unknown => {
    if (depth > 64) throw new TypeError("MessagePack nesting exceeds 64 levels");
    const marker = number(1);
    if (marker <= 0x7f) return marker;
    if (marker >= 0xe0) return marker - 256;
    if ((marker & 0xe0) === 0xa0) return decoder.decode(take(marker & 31));
    if ((marker & 0xf0) === 0x90) return array(marker & 15, depth);
    if ((marker & 0xf0) === 0x80) return map(marker & 15, depth);
    switch (marker) {
      case 0xc0: return null;
      case 0xc2: return false;
      case 0xc3: return true;
      case 0xc4: return take(number(1));
      case 0xc5: return take(number(2));
      case 0xc6: return take(number(4));
      case 0xca: return number(4, false, true);
      case 0xcb: return number(8, false, true);
      case 0xcc: return number(1);
      case 0xcd: return number(2);
      case 0xce: return number(4);
      case 0xcf: return number(8);
      case 0xd0: return number(1, true);
      case 0xd1: return number(2, true);
      case 0xd2: return number(4, true);
      case 0xd3: return number(8, true);
      case 0xd9: return decoder.decode(take(number(1)));
      case 0xda: return decoder.decode(take(number(2)));
      case 0xdb: return decoder.decode(take(number(4)));
      case 0xdc: return array(number(2), depth);
      case 0xdd: return array(number(4), depth);
      case 0xde: return map(number(2), depth);
      case 0xdf: return map(number(4), depth);
      default: throw new TypeError(`Unsupported MessagePack marker: 0x${marker.toString(16)}`);
    }
  };
  const value = read(0);
  if (offset !== bytes.length) throw new TypeError("MessagePack frame contains trailing data");
  return value;
}

File: sdks/python/speechswitch/msgpack.py
"""MessagePack scalars, binary values and containers used by Fish's wire protocol.

No extension types, non-string map keys, duplicate keys or trailing data. Numeric
integers decoded from 64-bit wire forms must fit the canonical safe-integer range.
"""

import math
import struct
from speechswitch.validation import is_mapping, is_sequence

type Value = None | bool | int | float | str | bytes | list[Value] | dict[str, Value]


def encode(value: object) -> bytes:
    output = bytearray()

    def header(marker: int, size: int = 0, length: int = 0) -> None:
        output.append(marker)
        output.extend(length.to_bytes(size, "big", signed=False))

    def write(value: object, depth: int) -> None:
        if depth > 64:
            raise TypeError("MessagePack nesting exceeds 64 levels")
        if value is None or type(value) is bool:
            header(0xc0 if value is None else 0xc3 if value else 0xc2)
        elif isinstance(value, (int, float)):
            try:
                number = float(value)
            except OverflowError:
                raise TypeError("MessagePack numbers must be finite") from None
            if not math.isfinite(number):
                raise TypeError("MessagePack numbers must be finite")
            integer = int(value)
            if value == integer and 0 <= integer <= 0x7f:
                header(integer)
            elif value == integer and -32 <= integer < 0:
                header(integer + 256)
            elif value == integer and 0 <= integer <= 0xffffffff:
                size = 1 if integer <= 0xff else 2 if integer <= 0xffff else 4
                header({1: 0xcc, 2: 0xcd, 4: 0xce}[size], size, integer)
            elif value == integer and -0x80000000 <= integer < 0:
                header(0xd2)
                output.extend(integer.to_bytes(4, "big", signed=True))
            else:
                header(0xcb)
                # Match the canonical TypeScript number's double representation.
                # Schema @integer bounds separately reject unsafe integer fields.
                output.extend(struct.pack(">d", number))
        elif isinstance(value, (str, bytes)):
            data = value.encode("utf-8") if isinstance(value, str) else value
            size = len(data)
            if size > 0xffffffff:
                raise TypeError("MessagePack value exceeds 32-bit length")
            if isinstance(value, str) and size < 32:
                header(0xa0 | size)
            else:
                width = 1 if size <= 0xff else 2 if size <= 0xffff else 4
                marker = {1: 0xd9, 2: 0xda, 4: 0xdb} if isinstance(value, str) else {1: 0xc4, 2: 0xc5, 4: 0xc6}
                header(marker[width], width, size)
            output.extend(data)
        elif is_mapping(value) or is_sequence(value):
            size = len(value)
            if size > 0xffffffff:
                raise TypeError("MessagePack value exceeds 32-bit length")
            mapping = is_mapping(value)
            if size < 16:
                header((0x80 if mapping else 0x90) | size)
            else:
                width = 2 if size <= 0xffff else 4
                header((0xde if width == 2 else 0xdf) if mapping else (0xdc if width == 2 else 0xdd), width, size)
            if is_mapping(value):
                for key, child in value.items():
                    if not isinstance(key, str):
                        raise TypeError("MessagePack map key is not a string")
                    write(key, depth + 1)
                    write(child, depth + 1)
            else:
                for child in value:
                    write(child, depth + 1)
        else:
            raise TypeError("Unsupported MessagePack value")

    write(value, 0)
    return bytes(output)


def decode(data: bytes) -> Value:
    offset = 0

    def take(size: int) -> bytes:
        nonlocal offset
        if size > len(data) - offset:
            raise TypeError("Truncated MessagePack value")
        result = data[offset:offset + size]
        offset += size
        return result

    def integer(size: int, signed: bool = False) -> int:
        value = int.from_bytes(take(size), "big", signed=signed)
        if size == 8 and abs(value) > 9007199254740991:
            raise TypeError("MessagePack integer exceeds the safe integer range")
        return value

    def array(size: int, depth: int) -> list[Value]:
        if size > len(data) - offset:
            raise TypeError("Truncated MessagePack value")
        return [read(depth + 1) for _ in range(size)]

    def mapping(size: int, depth: int) -> dict[str, Value]:
        if size > (len(data) - offset) // 2:
            raise TypeError("Truncated MessagePack value")
        result: dict[str, Value] = {}
        for _ in range(size):
            key = read(depth + 1)
            if not isinstance(key, str):
                raise TypeError("MessagePack map key is not a string")
            if key in result:
                raise TypeError("MessagePack map contains duplicate keys")
            result[key] = read(depth + 1)
        return result

    def read(depth: int) -> Value:
        if depth > 64:
            raise TypeError("MessagePack nesting exceeds 64 levels")
        marker = integer(1)
        if marker <= 0x7f:
            return marker
        if marker >= 0xe0:
            return marker - 256
        if marker & 0xe0 == 0xa0:
            return take(marker & 31).decode("utf-8")
        if marker & 0xf0 == 0x90:
            return array(marker & 15, depth)
        if marker & 0xf0 == 0x80:
            return mapping(marker & 15, depth)
        if marker == 0xc0:
            return None
        if marker in (0xc2, 0xc3):
            return marker == 0xc3
        if marker in (0xc4, 0xc5, 0xc6):
            return take(integer({0xc4: 1, 0xc5: 2, 0xc6: 4}[marker]))
        if marker in (0xca, 0xcb):
            value: float = struct.unpack(">f" if marker == 0xca else ">d", take(4 if marker == 0xca else 8))[0]
            return value
        if 0xcc <= marker <= 0xd3:
            return integer(1 << ((marker - 0xcc) % 4), marker >= 0xd0)
        if marker in (0xd9, 0xda, 0xdb):
            return take(integer({0xd9: 1, 0xda: 2, 0xdb: 4}[marker])).decode("utf-8")
        if marker in (0xdc, 0xdd):
            return array(integer(2 if marker == 0xdc else 4), depth)
        if marker in (0xde, 0xdf):
            return mapping(integer(2 if marker == 0xde else 4), depth)
        raise TypeError(f"Unsupported MessagePack marker: 0x{marker:x}")

    value = read(0)
    if offset != len(data):
        raise TypeError("MessagePack frame contains trailing data")
    return value
